### PR TITLE
Rand works in jit

### LINKED ISSUE #1957: JIT doesn't work with dropout
Fix LoadOps.RAND in the JIT

### PR DESCRIPTION
**Pull request by nimlgen**
Add a hook to regenerate rand buffers when jitted.

Closes #1957

### CODE REVIEW

File: tinygrad/lazy.py [hunk 1]
```diff
@@ -428,6 +428,8 @@ def _realize_empty(buffer: LazyBuffer) -> None:
 def _realize_rand(buffer: LazyBuffer) -> None:
   rng = np.random.default_rng(buffer.op.arg)
   buffer.realized = Device[buffer.device].buffer.fromCPU(rng.random(size=buffer.shape, dtype=np.float32).astype(dtype=buffer.dtype.np, copy=False), **buffer._device_extra_args()) # type: ignore
+  from tinygrad.jit import CacheCollector
+  CacheCollector.add(lambda args, vars, jit: args[0]._copyin(args[1].random(size=args[2], dtype=np.float32).astype(dtype=args[3], copy=False)), [buffer.realized, rng, buffer.shape, buffer.dtype.np], {})
```
Comment (geohot) [inline tinygrad/lazy.py:432]: That this line is duplicated shows me that an abstraction is wrong

Comment (nimlgen) [reply tinygrad/lazy.py]: You mean this random generation part or `CacheCollector.add`-staff?

**Comment by geohot** [comment]
That's not the solution. The abstraction is broken.

**Comment by geohot** [comment]
Sorry, removed

```
def _gen_rand(rng, shape, dt): return rng.random(size=shape, dtype=np.float32).astype(dtype=dt, copy=False)
def _realize_rand(buffer: LazyBuffer) -> None:
  rng = np.random.default_rng(buffer.op.arg)
  buffer.realized = Device[buffer.device].buffer.fromCPU(_gen_rand(rng, buffer.shape, buffer.dtype.np), **buffer._device_extra_args()) # type: ignore

  # Jit support
  from tinygrad.jit import CacheCollector
  CacheCollector.add(lambda args, vars, jit: args[0]._copyin(_gen_rand(*args[1:])), [buffer.realized, rng, buffer.shape, buffer.dtype.np], {})
  ```
  
  will add back and merge. The JIT API needs to be refactored with schedule, but it doesn't have to block this

**Comment by nimlgen** [comment]
Btw I have started with experimenting of rewriting JIT with schedule in #1952. Don't know how got abstractions there yet, but yeah I also thought of something like that. Would appreciate if you comment on `compiled schedule`, do we wanna have something like that?

**Comment by geohot** [comment]
Yea, we do want something like compiled schedule, but I think that type signature is wrong. I think this should be designed before it's coded, I want it to be a really big red diff.

**Comment by geohot** [comment]
Hmm, also this doesn't test that the RNG behavior matches in the JIT and not the JIT. I don't think it does.

**Comment by nimlgen** [comment]
Yeah, they do not match. Let's revert it, better to redesign jit. But since seed is changed in Tensor, I think even after redesign we gonna have some mismatch, no? Maybe set rng with a given seed in Tensor and just pass it?

**Pull request merged**